### DISCUSSION TITLE
initialData function on disabled queries

### DISCUSSION THREAD

**Discussion by TkDodo** [TanStack/query#1318] [Ideas]
This is a bit inspired by this issue:
https://github.com/tannerlinsley/react-query/issues/1296

I wanted to work around it by just passing the result from the bootstrap fetch as initialData to useQuery and just disable the query as long as the data is not yet loaded.

But it seems the initialData function is run _immediately_, even if the query is still `enabled: false`.

Would it make sense to run the initialData function only when the query is first enabled, so that we can make it dependent on a "previous" query?

Here would be an example (v3): https://codesandbox.io/s/dependent-initial-data-y39v6
Thoughts?

**Comment by tannerlinsley** [comment]
Yeah I think this is a good idea.

**Reply by TkDodo** [reply]
so I was trying to take a stab at this today, but I'm a bit lost tbh. I can see that currently, the initialData function is called from the constructor of the Query class, so I'd need to move that. But I'm having a hard time to figure out where to move it. Do you have maybe a code pointer to where a query would transition from `disabled` to `enabled`?

Another problem would maybe be that if you call e.g. `refetchQueries`, the initialData function would never be called, because the query is still `disabled`, but you would still fetch it I guess ...

**Reply by hayatae** [reply]
I've actually just run into this same issue. I'm waiting for my initial app state to load (determine whether or not the user is logged in) before enabling the query. I'd like to provide initialData only if the user is *not* logged in. I thought I could do an if statement in the initialData function based on the async loaded user state and simply not enable the query until the user state is known. Unfortunately, as described in this question, the initialData function is run immediately which means I'm attempting to make that determination too early.

Would absolutely love to see a fix for this!

**Reply by hayatae** [reply]
Actually just came up with one workaround for this issue which isn't great, but is functional:
Add your enabled boolean value to the `queryKey` array. That will cause the query to re-evaluate the initialData function once the enabled flag changes from false to true.

**Reply by hayatae** [reply]
Was doing some more testing and now I'm not sure if `initialData` is actually behaving correctly at all for `useInfiniteQuery`. Even though my `initialData` function is returning a value, my `queryFn` is still being executed immediately after. Nothing is calling `fetchMore`. I've tried setting `initialStale` explicitly to false and leaving it undefined. I'd expect the `queryFn` to not be invoked if `initialData` returns a value like `useQuery` does.

UPDATE:
This is apparently only a problem when `keepPreviousData: true,` is set on the infinite query. Without that setting, the `queryFn` is not invoked when `initialData` is provided. Is this intentional or a bug?

**Reply by TkDodo** [reply]
> Actually just came up with one workaround for this issue which isn't great, but is functional:
Add your enabled boolean value to the queryKey array. That will cause the query to re-evaluate the initialData function once the enabled flag changes from false to true.

right, this totally works and is not a bad idea. Here is the example with that version, and it works. Thanks!

https://codesandbox.io/s/dependent-initial-data-key-lcnnz

> Is this intentional or a bug?

It doesn't sound intentional to me. Can you show a reproducible example and then we can file an issue or so ...

**Reply by Fsss126** [reply]
Any progress on this one? Run into the same bummer with `initialData` being called just once when the query is disabled and not reinvoked when the query is enabled.

**Reply by TkDodo** [reply]
> Run into the same bummer with initialData being called just once when the query is disabled and not reinvoked when the query is enabled.

that’s per design. initial data function only runs when the query is created, and the query is created when `useQuery` is first invoked, even if its temporary disabled.
